### PR TITLE
Angular: EventEmitter issue with StrictTemplates set to true

### LINKED ISSUE #155: Angular: EventEmitter issue with StrictTemplates set to true
Using Angular v12, when setting strictTemplates: true in tsconfig I get this error

`error TS2339: Property 'subscribe' does not exist on type 'EventEmitter<SbgFileUploadChangeEvent>`

The HTML causing the issue looks something like this 
`<my-comp (fileChange)="doSomething()"</my-comp> `

Taking a look at my component during runtime I can see that subscribe is there (see image below)

Screenshot 2021-05-24 at 08 22 05

Also, If I change the import of EventEmitter in the generated d.ts file for my component from stencil-public-runtime to angular/core this error goes away. Obviously this is not a viable solution though.

The generated proxy is shown below:

```
@ProxyCmp({
    inputs: ['allowedFileTypes', 'disabled', 'fileUploadStatus', 'maxCombinedFileSize', 'maxFileSize', 'multiple', 'required', 'sbgFileUploadStatus'],
    methods: ['reset', 'getSbgFileStatus']
})
@Component({
    selector: 'sbg-file-upload',
    changeDetection: ChangeDetectionStrategy.OnPush,
    template: '<ng-content></ng-content>',
    inputs: ['allowedFileTypes', 'disabled', 'fileUploadStatus', 'maxCombinedFileSize', 'maxFileSize', 'multiple', 'required', 'sbgFileUploadStatus'],
    outputs: ['fileChange']
})
export class SbgFileUploadComponent {
    @Input() disabled: any;

    fileChange!: ISbgFileUpload['fileChange'];
    protected el: HTMLElement;

    constructor(c: ChangeDetectorRef, r: ElementRef, protected z: NgZone) {
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['fileChange']);
    }
}
```

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fixes #155

### CODE REVIEW

**Comment by reda-alaoui** [comment]
The import on '@stencil/core' does not work.
I give up.

**Comment by FrancoisDF** [comment]
This looks quite promising. Why do you keep the reference to the EventEmitter from stenciljs core? it exports a type that is not the same as the one expected from angular.
StencilJs EventEmmiter:
```
export interface EventEmitter<T = any> {
    emit: (data?: T) => CustomEvent<T>;
}
```
Angular/core
```
export declare interface EventEmitter<T> extends Subject<T> {
    new (isAsync?: boolean): EventEmitter<T>;
    emit(value?: T): void;
    subscribe(next?: (value: T) => void, error?: (error: any) => void, complete?: () => void): Subscription;
    subscribe(observerOrNext?: any, error?: any, complete?: any): Subscription;
}
```
Why do you keep the `${output.name} &` in the type of the proxy output? I tried based on your code without it and I'm not seeing any issues..

**Comment by reda-alaoui** [comment]
> Why do you keep the reference to the EventEmitter from stenciljs core?

I need a way to extract type X from Stenciljs EventEmitter<X> to use it as Y in Angular EventEmitter<Y> . I don’t know how to do it without referencing StencilJs EventEmitter.

**Comment by FrancoisDF** [comment]
What you have done with `StencilEventEmitterValueType` should be enough. 
The only thing I did was to remove the `${typename}` from the variable declaration:
```
   const typeName = `I${cmpMeta.componentClassName}['${output.method}']`;
    lines.push(`  ${output.name}!: EventEmitter<StencilEventEmitterValueType<${typeName}>>;`);
```
This keeps the type while using the `EventEmitter` from angular.
I do not get any build compilation even with ivy enabled and `skipLibCheck` and `skipDefaultLibCheck` both set to false.

**Comment by reda-alaoui** [comment]
Apps consuming StencilJs web-component are not expected to have any dependency on StencilJs itself. This is one of the pillars of StencilJs (no framework dependency), isn’t it?

In practice, when I use my web component (generated by the patched generator) from my angular application, it fails because the latter cannot resolve `@stencil/core`.

Maybe your consuming app has a direct dependency on stencil?

**Comment by reda-alaoui** [comment]
To be honest, I am not sure I understand what you do differently from me. Can you show the content generated by your patched generator?

**Comment by FrancoisDF** [comment]
What you have:
```
    lines.push(`  ${output.name}!: ${typeName} & EventEmitter<StencilEventEmitterValueType<${typeName}>>;`);
```
What I have:
```
    lines.push(`  ${output.name}!: EventEmitter<StencilEventEmitterValueType<${typeName}>>;`);
```
I was confused by the `${typeName} &`. I'm only considering the EventEmitter from angular.

File: packages/angular-output-target/src/output-angular.ts [hunk 1]
```diff
@@ -72,16 +72,20 @@ export function generateProxies(
 
   const imports = `/* tslint:disable */
 /* auto-generated angular directive proxies */
-import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, NgZone } from '@angular/core';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';
+import { EventEmitter as StencilEventEmitter } from '@stencil/core';
```
Comment (reda-alaoui) [inline packages/angular-output-target/src/output-angular.ts:76]: Ok I see. But there is still the reference to @stencil/core that is problematic, isn’t it?

Comment (FrancoisDF) [reply packages/angular-output-target/src/output-angular.ts:76]: What about using the EventEmitter part of the core generated library? It exposes a `stencil-public-runtime` type definition. So maybe generating something like that:
```
import { EventEmitter as StencilEventEmitter } from 'component-library/dist/types/stencil-public-runtime';
```

File: packages/angular-output-target/__tests__/output-angular.spec.ts [hunk 2]
```diff
@@ -19,10 +19,12 @@ describe('generateProxies', () => {
     expect(finalText).toEqual(
       `/* tslint:disable */
 /* auto-generated angular directive proxies */
-import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, NgZone } from '@angular/core';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';
+import { EventEmitter as StencilEventEmitter } from '@stencil/core';
 import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';
 
 import { Components } from 'component-library';
+type StencilEventEmitterValueType<Type> = Type extends StencilEventEmitter<infer X> ? X : never
```
Comment (FrancoisDF) [inline packages/angular-output-target/__tests__/output-angular.spec.ts:28]:
```ts
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';

import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';
import { EventEmitter as StencilEventEmitter } from 'component-library/dist/types/stencil-public-runtime';
import { Components } from 'component-library';
type StencilEventEmitterValueType<Type> = Type extends StencilEventEmitter<infer X> ? X : never
```

File: packages/angular-output-target/__tests__/output-angular.spec.ts [hunk 3]
```diff
@@ -37,12 +39,15 @@ import { Components } from 'component-library';
     expect(finalText).toEqual(
       `/* tslint:disable */
 /* auto-generated angular directive proxies */
-import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, NgZone } from '@angular/core';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';
+import { EventEmitter as StencilEventEmitter } from '@stencil/core';
 import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';
 
 import { Components } from '../../angular-output-target/dist/types/components';
+type StencilEventEmitterValueType<Type> = Type extends StencilEventEmitter<infer X> ? X : never
```
Comment (FrancoisDF) [inline packages/angular-output-target/__tests__/output-angular.spec.ts:48]:
```ts
/* auto-generated angular directive proxies */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';

import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';
import { EventEmitter as StencilEventEmitter } from '../../angular-output-target/dist/types/stencil-public-runtime';
import { Components } from '../../angular-output-target/dist/types/components';
type StencilEventEmitterValueType<Type> = Type extends StencilEventEmitter<infer X> ? X : never
```

File: packages/angular-output-target/src/output-angular.ts (hunk 1 above)
Comment (FrancoisDF) [inline packages/angular-output-target/src/output-angular.ts:76]: something like that could work
```ts
 ...
  const publicRuntimeFilePath = path.join(rootDir, distTypesDir, 'stencil-public-runtime');
  const publicRuntimeTypeFile = relativeImport(outputTarget.directivesProxyFile, publicRuntimeFilePath, '.d.ts');

  let imports = `/* tslint:disable */
/* auto-generated angular directive proxies */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, NgZone } from '@angular/core';

import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';\n`;

  if( !outputTarget.componentCorePackage){
    imports += `import { EventEmitter as StencilEventEmitter } from '${normalizePath(publicRuntimeTypeFile)}';`
  } else {
    imports += `import { EventEmitter as StencilEventEmitter } from '${normalizePath(path.join(outputTarget.componentCorePackage, distTypesDir, 'stencil-public-runtime'))}';`
  }
```

**Review by FrancoisDF — COMMENTED** [review]
I added few things to reference `stencil-public-runtime` available in the core components folder. I'm not fond of the `normalizePath(path.join...`, but that gives you a possible implementation.

File: packages/angular-output-target/src/output-angular.ts (hunk 1 above)
Comment (reda-alaoui) [reply packages/angular-output-target/src/output-angular.ts:76]: Sorry for my unresponsiveness, I am on holiday.
Why don't you create pull request based on this one to push it to the finish line? :)
